y/iterator.go

```diff
@@ -75,38 +75,61 @@ type Iterator interface {
 // MergeTowIterator is a specialized MergeIterator that only merge tow iterators.
 // It is an optimization for compaction.
 type MergeIterator struct {
+	smallerValid bool
+	biggerValid  bool
+	smallerKey   []byte
+	biggerKey    []byte
+	smallerVal   ValueStruct
+	biggerVal    ValueStruct
+
 	second  Iterator
 	smaller Iterator
 	bigger  Iterator
 	reverse bool
 }
 
 func (mt *MergeIterator) fix() {
-	if !mt.bigger.Valid() {
+	mt.resetSmaller()
+	if !mt.biggerValid {
 		return
 	}
 	var cmp int
-	for mt.smaller.Valid() {
-		cmp = CompareKeys(mt.smaller.Key(), mt.bigger.Key())
+	for mt.smallerValid {
```

Comment:
The `smaller` is an interface,  cannot be inlined.
We cache the children properties to avoid the interface function call.

RocksDB has a similar optimization:
https://github.com/facebook/rocksdb/blob/8bf555f487d1de84a4fb19cb97b9ae1a8dbebc60/table/iterator_wrapper.h#L22